### ISSUE TITLE
Settings overhaul tracker

### ISSUE THREAD

**Issue by fatcerberus** [Radnen/spherestudio#45]
Going to make a new issue thread for this, as it turns out this is going to be a fairly large undertaking.

Ignoring the issue of INI vs. XML for the time being because that's kind of a bikeshed debate.  The bigger issue, and one we should tackle now, is presets.  Presets are currently just that: A snapshot of all options at the time the preset was saved.  Loading a preset, as by using the config selector, may overwrite settings which seem like they should be global--such as "load last project on startup".

To fix this, there should be a distinction between configuration presets and settings.  The preset could store engine and project paths and activated plugins.  The global settings object would then have the rest of the options and apply to all presets.

Thoughts?

**Comment by Radnen** [comment]
Long post ahead; all good talking points.

My thought here is it should be perfectly fine to separate the two. We can go a bit further and try to tie the project settings on a per game basis, like .vcproj files. I mean, we ask ourselves what part of the settings ought to be committed into a repository.

BTW, I'm going to add all opened documents to a new kind of local project file. Every game would have their `game.sgm` file and then there will be a `.ssuser` filetype which does not get sent to the repo (contains user defaults for docked areas and opened files and settings relevant to just *that* project), and a `.ssproj` filetype which does get sent to the repo. That file would contain the build attributes, including things like arg handling because Sphere games actually *do* read the argument list from the cmd line. This is a functionality the old editor had that I had not added since almost nobody used that feature. But seeing how Sphere Studio is evolving, this might be something to add back in, especially for those who intend to make some serious games and need the added ability to debug the game via cmd-line arguments.

Then there should be global settings which can be stored like usual which contains things like 'open last project'. Then presets as well.

An issue I see is a conflict between the preset and a "repo-submitable" `.ssproj` file. The preset has engine setups, but the `.ssproj` file would be a preset sort of set up for just *that* project (and so all collaborators can see the same "build options").

Speaking of a build, I may even add some processors to it like jshint and minification. I mean, though these aren't web games I like to think there is a sense of  finalization in a minified JS file, saying, "this is for deployment, do not mess with it". This pipeline can be configured in the `.ssproj` file.

And just one last time, again, opened files are stored in the `.ssuser` file, and is not sent to the repo but still stored locally as a convenience file (for each project you want to see your last edited files).

**Comment by fatcerberus** [comment]
Hm, it's definitely a radical change.  I think for now we should take this one step at a time.  I'll get the ISettings stuff worked out.  What I think I'll do with that is change it so that each plugin gets passed an ISettings object that points to an INI file for just that plugin.  If the plugin needs to access a different INI file, it has to be done manually still, but for convenience it will get its own with no extra effort.  The main config will still be accessible through `PluginManager.IDE.Settings`, also.

I don't know that I want to get rid of game.sgm entirely though.  That to me has always been one of Sphere's main strengths: The game is runnable by the engine at any stage of development.  When you want to distribute, you just make an SPK (which mostly renders minification redundant since the scripts are then compressed).  But the .ssproj does have its benefits, no denying that.

I guess we'll see what happens down the road. :smiley:

**Comment by Radnen** [comment]
No, the `game.sgm` won't be removed entirely, but it doesn't have to be committed to the repo if the game settings are in the `.ssproj` file. So, it's a file made on project build. So yeah, I'm going to introduce a build pipeline to SphereStudio.

Once that concept is done I'm going to take the idea of processors and add them later. Things like minification, sprite sheeting (y'know, a process that puts your individual sprite images into a single spritesheet), image scaling, font generation, etc. It's almost endless.

I'll definitely be taking a few pages from XNA. But first thing's first is to work on the new `.ssuser` and `.ssproj` files and then the integration of a build/pipeline. This is going to take a while to do.

**Comment by fatcerberus** [comment]
I like this .ssproj idea, I'm fully behind it.  The only stipulation I would push for is not to give up the ability for no-build test runs--clicking Test Game shouldn't have to run a build step every time.  At least not by default.

This will be a great thing to have once Sphere has a NuGet-like package system as well--the .ssproj can specify the packages to download.

**Comment by Radnen** [comment]
Exactly. It won't rebuild every time. But some things may need reprocessing once I get to that step. It'll definitely help with the NuGet like packaging though.

BTW, I Just want to mention that these processors are either plugins for Sphere Studio *or* script files that look like this:

```javascript
function game(input_file, output_file) {
    var file = LoadFile(input_file);
    /* do some processing */
    SaveFile(file, output_file);
    return 1; // or anything 'truthy'
}
```

Returning 0/falsy means it failed. Load/Save file are just placeholders for whatever method you use. The only issue here is these scripts can't be sandboxed depending where you build to/from. Thesescripts may be invoked with Jurassic built into the editor. It's so good I may even add a plugin API for it so the script editors could get back syntax errors - wouldn't that be neat? Gha! I've got too many ideas!!

**Comment by fatcerberus** [comment]
New Configuration Manager with more intuitive preset editing:
https://github.com/Radnen/spherestudio/commit/93e1502590c2a10f431eb5c5460b22841dcba464

**Comment by fatcerberus** [comment]
Going to close this since the settings overhaul is in master now in the form of the new Settings Center.
